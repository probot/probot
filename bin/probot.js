#!/usr/bin/env node

import "../lib/bin/probot.js";
