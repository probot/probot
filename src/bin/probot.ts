import semver from "semver";
import program from "commander";
import updateNotifier from "update-notifier";

require("dotenv").config();

const pkg = require("../../package");

if (!semver.satisfies(process.version, pkg.engines.node)) {
  console.log(
    `Node.js version ${pkg.engines.node} is required. You have ${process.version}.`
  );
  process.exit(1);
}

if (pkg.version !== "0.0.0-development") {
  updateNotifier({ pkg }).notify();
}

program
  .version(pkg.version)
  .usage("<command> [options]")
  .command("run", "run the bot")
  .command("receive", "Receive a single event and payload")
  .on("command:*", (cmd) => {
    if (!program.commands.find((c) => c._name == cmd[0])) {
      console.error(`Invalid command: ${program.args.join(" ")}\n`);
      program.outputHelp();
      process.exit(1);
    }
  });

program.parse(process.argv);
