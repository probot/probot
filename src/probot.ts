import LRUCache from "lru-cache";
import { Logger } from "pino";
import { EmitterWebhookEvent as WebhookEvent } from "@octokit/webhooks";

import { aliasLog } from "./helpers/alias-log";
import { auth } from "./auth";
import { getLog } from "./helpers/get-log";
import { getProbotOctokitWithDefaults } from "./octokit/get-probot-octokit-with-defaults";
import { getWebhooks } from "./octokit/get-webhooks";
import { ProbotOctokit } from "./octokit/probot-octokit";
import { VERSION } from "./version";
import {
  ApplicationFunction,
  DeprecatedLogger,
  Options,
  ProbotWebhooks,
  State,
} from "./types";

export type Constructor<T> = new (...args: any[]) => T;

export class Probot {
  static version = VERSION;
  static defaults<S extends Constructor<any>>(this: S, defaults: Options) {
    const ProbotWithDefaults = class extends this {
      constructor(...args: any[]) {
        const options = args[0] || {};
        super(Object.assign({}, defaults, options));
      }
    };

    return ProbotWithDefaults;
  }

  public webhooks: ProbotWebhooks;
  public log: DeprecatedLogger;
  public version: String;
  public on: ProbotWebhooks["on"];
  public onAny: ProbotWebhooks["onAny"];
  public onError: ProbotWebhooks["onError"];
  public auth: (
    installationId?: number,
    log?: Logger
  ) => Promise<InstanceType<typeof ProbotOctokit>>;

  private state: State;

  constructor(options: Options = {}) {
    options.secret = options.secret || "development";

    let level = options.logLevel;
    const logMessageKey = options.logMessageKey;

    this.log = aliasLog(options.log || getLog({ level, logMessageKey }));

    // TODO: support redis backend for access token cache if `options.redisConfig`
    const cache = new LRUCache<number, string>({
      // cache max. 15000 tokens, that will use less than 10mb memory
      max: 15000,
      // Cache for 1 minute less than GitHub expiry
      maxAge: 1000 * 60 * 59,
    });

    const Octokit = getProbotOctokitWithDefaults({
      githubToken: options.githubToken,
      Octokit: options.Octokit || ProbotOctokit,
      appId: Number(options.appId),
      privateKey: options.privateKey,
      cache,
      log: this.log,
      redisConfig: options.redisConfig,
      baseUrl: options.baseUrl,
    });
    const octokit = new Octokit();

    this.state = {
      cache,
      githubToken: options.githubToken,
      log: this.log,
      Octokit,
      octokit,
      webhooks: {
        secret: options.secret,
      },
      appId: Number(options.appId),
      privateKey: options.privateKey,
      host: options.host,
      port: options.port,
    };

    this.auth = auth.bind(null, this.state);

    this.webhooks = getWebhooks(this.state);

    this.on = this.webhooks.on;
    this.onAny = this.webhooks.onAny;
    this.onError = this.webhooks.onError;

    this.version = VERSION;
  }

  public receive(event: WebhookEvent) {
    this.log.debug({ event }, "Webhook received");
    return this.webhooks.receive(event);
  }

  public async load(appFn: ApplicationFunction | ApplicationFunction[]) {
    if (Array.isArray(appFn)) {
      for (const fn of appFn) {
        await this.load(fn);
      }
      return;
    }

    return appFn(this, {});
  }
}
