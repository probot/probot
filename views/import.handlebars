<!DOCTYPE html>
<html lang="en" class="height-full" data-color-mode="auto" data-light-theme="light" data-dark-theme="dark">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Import {{#if pkg.name }}{{ pkg.name }}{{else}}Your App{{/if}} | built with Probot</title>
    <link rel="icon" href="/probot/static/probot-head.png">
    <link rel="stylesheet" href="/probot/static/primer.css">
</head>

<body class="bg-gray-light">
    <div class="d-flex flex-column flex-justify-center flex-items-center text-center height-full py-6">
        <a href="/probot"><img src="/probot/static/robot.svg" alt="Probot Logo" width="100" class="mb-6"></a>
        <div class="box-shadow rounded-2 border p-6 bg-white">
            <h2>Use existing Github App</h2>
            <br>

            <h3>Step 1:</h3>
            <p class="d-block mt-2">
                Replace your app's Webhook URL with <br>
                <b>{{ WEBHOOK_PROXY_URL }}</b>
            </p>
            <a class="d-block mt-2" href="{{ GH_HOST }}/settings/apps" target="__blank" rel="noreferrer">
                You can do it here
            </a>

            <br>
            <h3>Step 2:</h3>
            <p class="mt-2">Fill out this form</p>
            <form onsubmit="return onSubmit(event) || false">
                <label class="d-block mt-2" for="appId">App Id</label>
                <input class="form-control width-full" type="text" required="true" id="appId" name="appId"><br>

                <label class="d-block mt-3" for="whs">Webhook secret (required!)</label>
                <input class="form-control width-full" type="password" required="true" id="whs" name="whs"><br>

                <label class="d-block mt-3" for="pem">Private Key</label>
                <input class="form-control width-full m-2" type="file" accept=".pem" required="true" id="pem"
                    name="pem">
                <br>

                <button class="btn btn-outline m-2" type="submit">Submit</button>
            </form>
        </div>

        <div class="mt-4">
            <h4 class="alt-h4 text-gray-light">Need help?</h4>
            <div class="d-flex flex-justify-center mt-2">
                <a href="https://probot.github.io/docs/" class="btn btn-outline mr-2">Documentation</a>
                <a href="https://probot-slackin.herokuapp.com/" class="btn btn-outline">Chat on Slack</a>
            </div>
        </div>
    </div>
    <script>
        function onSubmit(e) {
            e.preventDefault();

            const idEl = document.getElementById('appId');
            const appId = idEl.value;


            const secretEl = document.getElementById('whs');
            const webhook_secret = secretEl.value;

            const fileEl = document.getElementById('pem');
            const file = fileEl.files[0];

            file.text().then((text) => fetch('', {
                method: 'POST',
                headers: { 'content-type': 'application/json' },
                body: JSON.stringify({ appId, pem: text, webhook_secret })
            })).then((r) => {
                if (r.ok) {
                    location.replace('/probot/success');
                }
            }).catch((e) => alert(e));
            return false;
        }
    </script>
</body>

</html>