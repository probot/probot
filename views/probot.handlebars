<!DOCTYPE html>
<html lang="en" class="height-full" data-color-mode="auto" data-light-theme="light" data-dark-theme="dark">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>{{#if name }}{{ name }}{{else}}Your App{{/if}} | built with Probot</title>
    <link rel="icon" href="/probot/static/probot-head.png">
    <link rel="stylesheet" href="/probot/static/primer.css">
  </head>
  <body class="height-full bg-gray-light">
    <div class="d-flex flex-column flex-justify-center flex-items-center text-center height-full">
      <img src="/probot/static/robot.svg" alt="Probot Logo" width="100" class="mb-6">
      <div class="box-shadow rounded-2 border p-6 bg-white">
        <h1>
          Welcome to {{#if name }}{{ name }}{{else}}your Probot App{{/if}}
          {{#if version}}
            <span class="Label Label--outline v-align-middle ml-2 text-gray-light">v{{ version }}</span>
          {{/if}}
        </h1>

        {{#if description }}
          <p>{{ description }}</p>
        {{else}}
          <p>This bot was built using <a href="https://github.com/probot/probot">Probot</a>, a framework for building GitHub Apps.</p>
        {{/if}}
      </div>

      <div class="mt-4">
        <h4 class="alt-h4 text-gray-light">Need help?</h4>
        <div class="d-flex flex-justify-center mt-2">
          <a href="https://probot.github.io/docs/" class="btn btn-outline mr-2">Documentation</a>
          <a href="https://probot-slackin.herokuapp.com/" class="btn btn-outline">Chat on Slack</a>
        </div>
      </div>
    </div>
  </body>
</html>
