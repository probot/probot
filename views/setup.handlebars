<!DOCTYPE html>
<html lang="en" class="height-full" data-color-mode="auto" data-light-theme="light" data-dark-theme="dark">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Setup {{#if pkg.name }}{{ pkg.name }}{{else}}Your App{{/if}} | built with Probot</title>
    <link rel="icon" href="/probot/static/probot-head.png">
    <link rel="stylesheet" href="/probot/static/primer.css">
  </head>
  <body class="height-full bg-gray-light">
    <div class="d-flex flex-column flex-justify-center flex-items-center text-center height-full">
      <img src="/probot/static/robot.svg" alt="Probot Logo" width="100" class="mb-6">
      <div class="box-shadow rounded-2 border p-6 bg-white">
        <h1>
          Welcome to {{#if pkg.name }}{{ pkg.name }}{{else}}your Probot App{{/if}}
          {{#if version}}
            <span class="Label Label--outline v-align-middle ml-2 text-gray-light">v{{ pkg.version }}</span>
          {{/if}}
        </h1>

        {{#if pkg.description }}
          <p>{{ pkg.description }}</p>
        {{else}}
          <p>This app was built using <a href="https://github.com/probot/probot">Probot</a>, a framework for building GitHub Apps.</p>
        {{/if}}

        <div class="text-left mt-6">
          <h2 class="alt-h3 mb-2">Getting Started</h2>

          <p>To start building a GitHub App, you'll need to register a new app on GitHub.</p>
          <br>

          <form action="{{ createAppUrl }}" method="post" target="_blank" class="d-flex flex-items-center">
            <button class="btn btn-outline" name="manifest" id="manifest" value='{{ manifest }}' >Register GitHub App</button>
            <a href="/probot/import" class="ml-2">or use an existing Github App</a>
          </form>
        </div>
      </div>

      <div class="mt-4">
        <h4 class="alt-h4 text-gray-light">Need help?</h4>
        <div class="d-flex flex-justify-center mt-2">
          <a href="https://probot.github.io/docs/" class="btn btn-outline mr-2">Documentation</a>
          <a href="https://probot-slackin.herokuapp.com/" class="btn btn-outline">Chat on Slack</a>
        </div>
      </div>
    </div>
  </body>
</html>
