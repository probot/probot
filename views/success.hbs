<!DOCTYPE html>
<html lang="en" class="height-full" data-color-mode="auto" data-light-theme="light" data-dark-theme="dark">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Setup {{#if pkg.name }}{{ pkg.name }}{{else}}Your App{{/if}} | built with Probot</title>
    <link rel="icon" href="/probot/static/probot-head.png">
    <link rel="stylesheet" href="/probot/static/primer.css">
  </head>
  <body class="height-full bg-gray-light">
    <div class="d-flex flex-column flex-justify-center flex-items-center text-center height-full">
      <img src="/probot/static/robot.svg" alt="Probot Logo" width="100" class="mb-6">
      <div class="box-shadow rounded-2 border p-6 bg-white">
        <div class="text-center">
          <h1 class="alt-h3 mb-2">Congrats! You have successfully installed your app!
            <br>
            Checkout <a href="https://probot.github.io/docs/webhooks/">Receiving webhooks</a> and <a href="https://probot.github.io/docs/github-api/">Interacting with GitHub</a> to learn more!</h1>
        </div>
      </div>

      <div class="mt-4">
        <h4 class="alt-h4 text-gray-light">Need help?</h4>
        <div class="d-flex flex-justify-center mt-2">
          <a href="https://probot.github.io/docs/" class="btn btn-outline mr-2">Documentation</a>
          <a href="https://probot-slackin.herokuapp.com/" class="btn btn-outline">Chat on Slack</a>
        </div>
      </div>
    </div>
  </body>
</html>
