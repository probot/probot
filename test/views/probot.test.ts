import { describe, expect, test } from "vitest";

import { probotView } from "../../src/views/probot";

describe("probotView", () => {
  test("not providing parameters", () => {
    expect(probotView({})).toMatchSnapshot();
  });

  test("providing name", () => {
    expect(
      probotView({
        name: "My App",
      }),
    ).toMatchSnapshot();
  });

  test("providing description", () => {
    expect(
      probotView({
        description: "My App with Probot",
      }),
    ).toMatchSnapshot();
  });

  test("providing description", () => {
    expect(
      probotView({
        version: "1.0.0",
      }),
    ).toMatchSnapshot();
  });
});
