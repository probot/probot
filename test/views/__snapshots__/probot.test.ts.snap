// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`probotView > not providing parameters 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\" class=\\"height-full\\" data-color-mode=\\"auto\\" data-light-theme=\\"light\\" data-dark-theme=\\"dark\\">
  <head>
    <meta charset=\\"UTF-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
    <title>Your App | built with Probot</title>
    <link rel=\\"icon\\" href=\\"/probot/static/probot-head.png\\">
    <link rel=\\"stylesheet\\" href=\\"/probot/static/primer.css\\">
  </head>
  <body class=\\"height-full bg-gray-light\\">
    <div class=\\"d-flex flex-column flex-justify-center flex-items-center text-center height-full\\">
      <img src=\\"/probot/static/robot.svg\\" alt=\\"Probot Logo\\" width=\\"100\\" class=\\"mb-6\\">
      <div class=\\"box-shadow rounded-2 border p-6 bg-white\\">
        <h1>
          Welcome to your Probot App
        </h1>

          <p>This bot was built using <a href=\\"https://github.com/probot/probot\\">Probot</a>, a framework for building GitHub Apps.</p>
      </div>

      <div class=\\"mt-4\\">
        <h4 class=\\"alt-h4 text-gray-light\\">Need help?</h4>
        <div class=\\"d-flex flex-justify-center mt-2\\">
          <a href=\\"https://probot.github.io/docs/\\" class=\\"btn btn-outline mr-2\\">Documentation</a>
          <a href=\\"https://probot-slackin.herokuapp.com/\\" class=\\"btn btn-outline\\">Chat on Slack</a>
        </div>
      </div>
    </div>
  </body>
</html>"
`;

exports[`probotView > providing description 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\" class=\\"height-full\\" data-color-mode=\\"auto\\" data-light-theme=\\"light\\" data-dark-theme=\\"dark\\">
  <head>
    <meta charset=\\"UTF-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
    <title>Your App | built with Probot</title>
    <link rel=\\"icon\\" href=\\"/probot/static/probot-head.png\\">
    <link rel=\\"stylesheet\\" href=\\"/probot/static/primer.css\\">
  </head>
  <body class=\\"height-full bg-gray-light\\">
    <div class=\\"d-flex flex-column flex-justify-center flex-items-center text-center height-full\\">
      <img src=\\"/probot/static/robot.svg\\" alt=\\"Probot Logo\\" width=\\"100\\" class=\\"mb-6\\">
      <div class=\\"box-shadow rounded-2 border p-6 bg-white\\">
        <h1>
          Welcome to your Probot App
        </h1>

          <p>My App with Probot</p>
      </div>

      <div class=\\"mt-4\\">
        <h4 class=\\"alt-h4 text-gray-light\\">Need help?</h4>
        <div class=\\"d-flex flex-justify-center mt-2\\">
          <a href=\\"https://probot.github.io/docs/\\" class=\\"btn btn-outline mr-2\\">Documentation</a>
          <a href=\\"https://probot-slackin.herokuapp.com/\\" class=\\"btn btn-outline\\">Chat on Slack</a>
        </div>
      </div>
    </div>
  </body>
</html>"
`;

exports[`probotView > providing description 2`] = `
"<!DOCTYPE html>
<html lang=\\"en\\" class=\\"height-full\\" data-color-mode=\\"auto\\" data-light-theme=\\"light\\" data-dark-theme=\\"dark\\">
  <head>
    <meta charset=\\"UTF-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
    <title>Your App | built with Probot</title>
    <link rel=\\"icon\\" href=\\"/probot/static/probot-head.png\\">
    <link rel=\\"stylesheet\\" href=\\"/probot/static/primer.css\\">
  </head>
  <body class=\\"height-full bg-gray-light\\">
    <div class=\\"d-flex flex-column flex-justify-center flex-items-center text-center height-full\\">
      <img src=\\"/probot/static/robot.svg\\" alt=\\"Probot Logo\\" width=\\"100\\" class=\\"mb-6\\">
      <div class=\\"box-shadow rounded-2 border p-6 bg-white\\">
        <h1>
          Welcome to your Probot App
            <span class=\\"Label Label--outline v-align-middle ml-2 text-gray-light\\">v1.0.0</span>
        </h1>

          <p>This bot was built using <a href=\\"https://github.com/probot/probot\\">Probot</a>, a framework for building GitHub Apps.</p>
      </div>

      <div class=\\"mt-4\\">
        <h4 class=\\"alt-h4 text-gray-light\\">Need help?</h4>
        <div class=\\"d-flex flex-justify-center mt-2\\">
          <a href=\\"https://probot.github.io/docs/\\" class=\\"btn btn-outline mr-2\\">Documentation</a>
          <a href=\\"https://probot-slackin.herokuapp.com/\\" class=\\"btn btn-outline\\">Chat on Slack</a>
        </div>
      </div>
    </div>
  </body>
</html>"
`;

exports[`probotView > providing name 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\" class=\\"height-full\\" data-color-mode=\\"auto\\" data-light-theme=\\"light\\" data-dark-theme=\\"dark\\">
  <head>
    <meta charset=\\"UTF-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
    <title>My App | built with Probot</title>
    <link rel=\\"icon\\" href=\\"/probot/static/probot-head.png\\">
    <link rel=\\"stylesheet\\" href=\\"/probot/static/primer.css\\">
  </head>
  <body class=\\"height-full bg-gray-light\\">
    <div class=\\"d-flex flex-column flex-justify-center flex-items-center text-center height-full\\">
      <img src=\\"/probot/static/robot.svg\\" alt=\\"Probot Logo\\" width=\\"100\\" class=\\"mb-6\\">
      <div class=\\"box-shadow rounded-2 border p-6 bg-white\\">
        <h1>
          Welcome to My App
        </h1>

          <p>This bot was built using <a href=\\"https://github.com/probot/probot\\">Probot</a>, a framework for building GitHub Apps.</p>
      </div>

      <div class=\\"mt-4\\">
        <h4 class=\\"alt-h4 text-gray-light\\">Need help?</h4>
        <div class=\\"d-flex flex-justify-center mt-2\\">
          <a href=\\"https://probot.github.io/docs/\\" class=\\"btn btn-outline mr-2\\">Documentation</a>
          <a href=\\"https://probot-slackin.herokuapp.com/\\" class=\\"btn btn-outline\\">Chat on Slack</a>
        </div>
      </div>
    </div>
  </body>
</html>"
`;
